Reliability: Introduce local vs store timezone difference tracks

ThomazFB opened the pull request:
Summary
==========
Fix issue #9144 by introducing a track event triggered whenever the device's local timezone configuration differs from the store timezone configuration. 

The purpose of this track is to allow us to evaluate how often our users see analytics inside the app at a different timezone from the store. Since the app uses the local device timezone whenever the store is configured differently, the analytics values and numbers between the wp-admin and app stats can significantly differ.

How to Test
==========
### Scenario 1
1. Open the app with a store configured in a different timezone from the device you're testing
2. Verify that the `DASHBOARD_STORE_TIMEZONE_DIFFER_FROM_DEVICE` event is triggered as soon as the `My Store` section is loaded
3. Verify that the same event contains the `store_timezone` and `local_timezone` carrying the expected values from each
4. Now, without reopening the app, go to the App More menu and switch the store to another one with a different timezone from the local
5. Verify that again the `DASHBOARD_STORE_TIMEZONE_DIFFER_FROM_DEVICE` event is triggered, containing the same `local_timezone` as before, and that `store_timezone` follows the value of the new selected store

### Scenario 2
1. Open the app with a store configured with the same timezone from the device you're testing
2. Verify that the `DASHBOARD_STORE_TIMEZONE_DIFFER_FROM_DEVICE` is NOT triggered


Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

ThomazFB committed:
Add Analytics Event for different timezone between store and device

ThomazFB committed:
Observe site changes inside the MyStoreViewModel

ThomazFB committed:
Add filter condition to only trigger the track event when local timezone is different from the store one

ThomazFB committed:
Extract store timezone observation from MyStoreViewModel init

ThomazFB committed:
Replace onEach with collect operation to correctly start the flow observation

ThomazFB committed:
Define TimeZoneExt.kt file

ThomazFB committed:
Implement offsetInHours property in TimeZoneExt.kt

ThomazFB committed:
Adjust Store timezone observation to use offsetInHours

ThomazFB committed:
Create local value for local timezone offset

ThomazFB committed:
Configure parameter keys for store and local timezones

ThomazFB committed:
Configure Analytics parameters

ThomazFB committed:
Fix lint issues

ThomazFB committed:
Merge branch 'trunk' into fix/sync-analytics-timezones

atorresveiga commented:
Great work, Thomaz! The code looks awesome 🙌 .
I think I found one small issue, though.
When testing the tracking of the event, I found out that we are sending the event twice, here is a video:

https://github.com/woocommerce/woocommerce-android/assets/18119390/11114d29-2ec8-4ead-9c85-7eb3bb771d9b

I'm not sure why this is happening. I think that we restart the activity when we switch stores, and this could trigger the event again 🤔 

This is what I see on the Logcat.

```
🔵 Tracked: dashboard_store_timezone_differ_from_device, Properties: {"store_timezone":"0","local_timezone":"-3","blog_id":214253715,"is_wpcom_store":true,"is_debug":true}
🔵 Tracked: dashboard_store_timezone_differ_from_device, Properties: {"store_timezone":"0","local_timezone":"-3","blog_id":186855689,"is_wpcom_store":false,"is_debug":true}
```

atorresveiga commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/mystore/MyStoreViewModel.kt:
```diff
@@ -323,6 +329,26 @@ class MyStoreViewModel @Inject constructor(
         }
     }
 
+    private fun observeStoreTimezoneChanges() {
+        viewModelScope.launch {
+            val localTimeZoneOffset = TimeZone.getDefault().offsetInHours.toString()
+
+            selectedSite.observe()
+                .filterNotNull()
+                .distinctUntilChanged { old, new -> new.id == old.id }
+                .filter { it.timezone != localTimeZoneOffset }
+                .collect {
+                    analyticsTrackerWrapper.track(
+                        stat = DASHBOARD_STORE_TIMEZONE_DIFFER_FROM_DEVICE,
+                        properties = mapOf(
+                            AnalyticsTracker.KEY_STORE_TIMEZONE to it.timezone,
+                            AnalyticsTracker.KEY_LOCAL_TIMEZONE to localTimeZoneOffset
+                        )
+                    )
+                }
+        }
+    }
```
Np: One small np is that we can add a unit test to this logic. We can always address this on a different PR, though.

ThomazFB replied:
Indeed! Thanks for the suggestion! I introduced some unit tests, but defining a `TimezoneProvider` was necessary to allow us to mock the device Timezone acquisition since it used a static field value provided by the environment. Let me know what you think!

atorresveiga commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/mystore/MyStoreViewModel.kt:
```diff
@@ -323,6 +329,26 @@ class MyStoreViewModel @Inject constructor(
         }
     }
 
+    private fun observeStoreTimezoneChanges() {
```
Nice!! 🙌

atorresveiga commented:
> Open the app with a store configured in a different timezone from the device you're testing

On scenario 2 you mean here:  Open the app with a store configured in <s>a different</s> **_the same_** timezone from the device you're testing, right?

ThomazFB committed:
Create empty test scenarios for analytics timezone event

ThomazFB committed:
Define TimezoneProvider

ThomazFB committed:
Use injected timezoneProvider inside MyStoreViewModel

ThomazFB committed:
Make TimezoneProvider injectable

ThomazFB committed:
Introduce base test scenario for timezone tracks in MyStoreViewModelTest

ThomazFB committed:
Finish MyStoreViewModelTest scenarios for timezone tracks

ThomazFB committed:
Fix lint issues

ThomazFB committed:
Update MyStoreViewModel timezone track to do it once when the ViewModel starts

ThomazFB committed:
Fix Failing MyStoreViewModelTest

ThomazFB committed:
Add null store scenario in MyStoreViewModelTest

ThomazFB commented:
A very interesting bug you found, @atorresveiga. Thank you so much! 🙇  After some investigations, I can say that this bug is a deal breaker for us. The repetition you observed would be fine since we want to see this from a unique user perspective. But the problem is the fact this repetition comes along with incorrect data! The provided Site flow seems to emit the new store before the `AnalyticsTracker` is updated with it. Check this out:

<img width="1210" alt="Screenshot 2023-06-01 at 06 17 19" src="https://github.com/woocommerce/woocommerce-android/assets/5920403/f4d0ad83-d6ee-4c5c-8667-b746b8846fd6">

When I change the store, it immediately emits a new site with the timezone of the new one and the `blog_id` of the old, and once the site switching ends, it emits once again, but now with the correct timezone. This will mix up the data and create trouble for us. 

With this problem, I decided to abandon the `selectedSite.observe()` usage and take advantage of the fact that the `MyStoreViewModel` is re-initialized every time the store changes. Now the track will happen only when the app starts or the store changes. WDYT? I updated the implementation in https://github.com/woocommerce/woocommerce-android/pull/9139/commits/71391bc4edf3a61952fbf458f22b59dec1129708 and added a third test in https://github.com/woocommerce/woocommerce-android/pull/9139/commits/0ab9b710bef70d454accefb4335e7290e44cb9be to make sure we won't be doing anything if the site object doesn't exist yet.

The bug should be fixed now!

ThomazFB commented:
>On scenario 2 you mean here: Open the app with a store configured in a different the same timezone from the device you're testing, right?

Oops! You're correct! 😅 The test description is updated now.

atorresveiga commented on WooCommerce/src/test/kotlin/com/woocommerce/android/ui/mystore/MyStoreViewModelTest.kt:
```diff
@@ -417,6 +423,87 @@ class MyStoreViewModelTest : BaseUnitTest() {
+        whenViewModelIsCreated()
+
+        // Then
+        verify(analyticsTrackerWrapper, times(0)).track(
```
np: we can use `never()` here instead of `times(0)`

ThomazFB replied:
True! Thanks for the reminder! Fixed in https://github.com/woocommerce/woocommerce-android/pull/9139/commits/25b7b5661b4e5b50b704d1b8b6cbe5f1928082e9

atorresveiga commented on WooCommerce/src/test/kotlin/com/woocommerce/android/ui/mystore/MyStoreViewModelTest.kt:
```diff
@@ -417,6 +423,87 @@ class MyStoreViewModelTest : BaseUnitTest() {
+        whenViewModelIsCreated()
+
+        // Then
+        verify(analyticsTrackerWrapper, times(0)).track(
```
np: we can use `never()` here instead of `times(0)`

ThomazFB replied:
True! Thanks for the reminder! Fixed in https://github.com/woocommerce/woocommerce-android/pull/9139/commits/25b7b5661b4e5b50b704d1b8b6cbe5f1928082e9

atorresveiga reviewed (approved):
Great work @ThomazFB! 🙌  Thanks for fixing the double tracking issue and unit testing the code 👌

LGTM! :shipit:

atorresveiga commented:
What do you think if we track this event only once per merchant store? My point is that right now, we will track this event every time a merchant opens the my store screen, but we only need to know how often the device time zone differs from the store time zone. Maybe we can add a flag to the device preferences with the key `device-timezone|store-timezone|blog-id` and check this flag before tracking the event (to track the event only once).
WDYT about this idea? 
If you agree, we can create a new task for this.

ThomazFB committed:
Remove unnecessary stubbing from MyStoreViewModelTest

ThomazFB committed:
Replace times(0) with never inside verify calls

ThomazFB commented:
>What do you think if we track this event only once per merchant store? My point is that right now, we will track this event every time a merchant opens the my store screen, but we only need to know how often the device time zone differs from the store time zone. Maybe we can add a flag to the device preferences with the key device-timezone|store-timezone|blog-id and check this flag before tracking the event (to track the event only once).
WDYT about this idea?
If you agree, we can create a new task for this.

I think it's a fair improvement, @atorresveiga. I'll create the tasks and implement this immediately before the code freeze. Thanks for the suggestion!
